features/events/daemon/src/main/java/org/opennms/netmgt/eventd/processor/expandable/EventTemplate.java

```diff
@@ -0,0 +1,137 @@
+/*******************************************************************************
+ * This file is part of OpenNMS(R).
+ *
+ * Copyright (C) 2016 The OpenNMS Group, Inc.
+ * OpenNMS(R) is Copyright (C) 1999-2016 The OpenNMS Group, Inc.
+ *
+ * OpenNMS(R) is a registered trademark of The OpenNMS Group, Inc.
+ *
+ * OpenNMS(R) is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published
+ * by the Free Software Foundation, either version 3 of the License,
+ * or (at your option) any later version.
+ *
+ * OpenNMS(R) is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with OpenNMS(R).  If not, see:
+ *      http://www.gnu.org/licenses/
+ *
+ * For more information contact:
+ *     OpenNMS(R) Licensing <license@opennms.org>
+ *     http://www.opennms.org/
+ *     http://www.opennms.com/
+ *******************************************************************************/
+
+package org.opennms.netmgt.eventd.processor.expandable;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+import org.opennms.netmgt.eventd.EventUtil;
+import org.opennms.netmgt.eventd.ExpandableParameterResolverRegistry;
+import org.opennms.netmgt.xml.event.Event;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.Lists;
+
+/**
+ * The {@link EventTemplate} represents the event input, containing place-holders, e.g. 'nodeLabel',
+ * and support to expand the template.
+ *
+ * For each placeholder in the {@link EventTemplate#input} a {@link ExpandableToken} is created.
+ * The expandable state of the {@link EventTemplate} is the expanded state of all tokens ({@link EventTemplate#tokens}.
+ */
+public class EventTemplate implements ExpandableToken {
+
+    private static final Logger LOG = LoggerFactory.getLogger(EventTemplate.class);
+
+    private static final char PERCENT = '%';
+
+    private static final String WHITESPACE_PATTERN = ".*\\s(?s).*";
+
+    private final String input;
+
+    private final List<ExpandableToken> tokens = Lists.newArrayList();
+
+    private final EventUtil eventUtil;
+
+    private final ExpandableParameterResolverRegistry parameterResolverRegistry;
+
+    public EventTemplate(String input, EventUtil eventUtil) {
+        this.input = Objects.requireNonNull(input);
+        this.eventUtil = Objects.requireNonNull(eventUtil);
+        this.parameterResolverRegistry = new ExpandableParameterResolverRegistry();
```

Comment:
Can we re-use the `ExpandableParameterResolverRegistry` instead of re-creating it for every `EventTemplate`. We could even pull it from the `EventUtil` instance that's being passed instead of passing it as another argument here.